File: numeric/max/max.cpp
#include <print>
#include <cstdlib>
#include <vector>

template<typename T, std::input_iterator iter>
int max(iter begin, iter end)
{
    T current = 0;
    for (iter it = begin; it != end; ++it)
    {
        if (*it > current)
        {
            current = *it;
        }
    }
    return current;
}


int main(int argc, char *argv[])
{
    std::vector<int> values{};

    if (argc > 1)
    {
        for (int i = 0; i < argc - 1; ++i)
        {
            values.push_back(std::stoi(argv[i + 1]));
        }
    }
    else
    {
        values.push_back(15);
        values.push_back(10);
    }

    int pmax = max<int>(values.begin(), values.end());

    std::println("values: {}", values);
    std::println("max: {}", pmax);
}

File: numeric/max/max.m
#import <Foundation/Foundation.h>
#include <stdlib.h>

int max(int values[], int count) {
    int current = 0;
    for (int i = 0; i < count; ++i) {
        if (values[i] > current) {
            current = values[i];
        }
    }
    return current;
}

int main(int argc, const char * argv[]) {
    int n;
    int* values;

    if (argc > 1)
    {
        values = (int*)malloc(sizeof(int) * argc - 1);
        for (int i = 0; i < argc - 1; ++i)
        {
            values[i] = atoi(argv[i + 1]);
        }
        n = argc - 1;
    }
    else
    {
        values = (int*)malloc(sizeof(int) * 2);
        values[0] = 15;
        values[1] = 10;
        n = 2;
    }

    int pmax = max(values, n);

    @autoreleasepool {
        NSLog(@"values:");
        for (int i = 0; i < n; ++i) {
            NSLog(@"%d", values[i]);
        }
        NSLog(@"gcd: %d", pmax);
    }

    free(values);

    return 0;
}
